modules/core/src/main/java/org/apache/fluo/core/async/AsyncConditionalWriter.java

```diff
@@ -84,21 +90,42 @@ public IterTask(Iterator<Result> iter, int permitsAcquired) {
 
   }
 
-  @Override
-  public ListenableFuture<Iterator<Result>> apply(Collection<ConditionalMutation> input) {
+  /*
+   * @Override public ListenableFuture<Iterator<Result>> apply(Collection<ConditionalMutation>
+   * input) { if (input.size() == 0) { return
+   * Futures.immediateFuture(Collections.<Result>emptyList().iterator()); }
+   * 
+   * semaphore.acquire(input.size()); Iterator<Result> iter = cw.write(input.iterator()); return
+   * les.submit(new IterTask(iter, input.size())); }
+   */
+
+  // @Override
+  public CompletableFuture<Iterator<Result>> apply(Collection<ConditionalMutation> input) {
+    CompletableFuture<Iterator<Result>> cfRes = new CompletableFuture<>();
     if (input.size() == 0) {
-      return Futures.immediateFuture(Collections.<Result>emptyList().iterator());
+      cfRes.complete(Collections.<Result>emptyList().iterator());
+      return cfRes;
     }
 
     semaphore.acquire(input.size());
     Iterator<Result> iter = cw.write(input.iterator());
-    return les.submit(new IterTask(iter, input.size()));
+    CompletableFuture.runAsync(() -> {
```

Comment:
Could inline and remove IterTask and do something like the following.

```java
    return CompletableFuture.supplyAsync(() -> {
      try {
        Builder<Result> imlb = ImmutableList.builder();
        while (iter.hasNext()) {
          Result result = iter.next();
          imlb.add(result);
        }
        return imlb.build().iterator();
      } finally {
        semaphore.release(permitsAcquired);
      }
    }, es);
```

Reply:
This change and the one above seem much cleaner, I'll make the appropriate changes.